lapack/gonum/dlangb.go

```diff
@@ -0,0 +1,100 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/internal/asm/f64"
+	"gonum.org/v1/gonum/lapack"
+)
+
+// Dlangb returns the given norm of an m×n band matrix with kl sub-diagonals and
+// ku super-diagonals.
+//
+// When norm is lapack.MaxColumnSum, the length of work must be at least n.
+func (impl Implementation) Dlangb(norm lapack.MatrixNorm, m, n, kl, ku int, ab []float64, ldab int, work []float64) float64 {
+	ncol := kl + 1 + ku
+	switch {
+	case norm != lapack.MaxAbs && norm != lapack.MaxRowSum && norm != lapack.MaxColumnSum && norm != lapack.Frobenius:
+		panic(badNorm)
+	case m < 0:
+		panic(mLT0)
+	case n < 0:
+		panic(nLT0)
+	case kl < 0:
+		panic(klLT0)
+	case ku < 0:
+		panic(kuLT0)
+	case ldab < ncol:
+		panic(badLdA)
+	}
+
+	// Quick return if possible.
+	if m == 0 || n == 0 {
+		return 0
+	}
+
+	switch {
+	case len(ab) < min(m, n+kl)*ldab:
+		panic(shortAB)
+	case len(work) < n && norm == lapack.MaxColumnSum:
+		panic(shortWork)
+	}
+
+	var value float64
+	switch norm {
+	case lapack.MaxAbs:
+		for i := 0; i < min(m, n+kl); i++ {
+			l := max(0, kl-i)
+			u := min(n+kl-i, ncol)
+			for _, aij := range ab[i*ldab+l : i*ldab+u] {
+				aij = math.Abs(aij)
+				if aij > value || math.IsNaN(aij) {
+					value = aij
+				}
+			}
+		}
+	case lapack.MaxRowSum:
+		for i := 0; i < min(m, n+kl); i++ {
+			l := max(0, kl-i)
+			u := min(n+kl-i, ncol)
+			sum := f64.L1Norm(ab[i*ldab+l : i*ldab+u])
+			if sum > value || math.IsNaN(sum) {
+				value = sum
+			}
+		}
+	case lapack.MaxColumnSum:
+		work = work[:n]
+		for j := range work {
+			work[j] = 0
+		}
+		for i := 0; i < min(m, n+kl); i++ {
+			l := max(0, kl-i)
+			u := min(n+kl-i, ncol)
+			for jb, aij := range ab[i*ldab+l : i*ldab+u] {
+				j := l + jb - kl + i
+				work[j] += math.Abs(aij)
+			}
+		}
```

Comment:
Can this be done easily with `f64.L1NormInc` directly into `work`?

Reply:
I'm sure it's doable: a formula for the top element (row and col in the band storage) in a column and a formula for the number of elements in a column ... at least `inc` would clearly be `ldab-1`. Uninviting but then work wouldn't be needed ...

Reply:
I've pushed a commit that does it. It's not that bad, the index calculations are pretty similar but still any idea for simplification is welcome. Now `work` is not needed, which makes Dlangb an outlier among the norm routines. I don't have benchmarks for what is faster: iterating with increment one or using assembly (on amd64) but I think we can keep this.

Reply:
I can't see any simplifications in that. It might be interesting to see what the performance differences are though.

Reply:
Hmm, there might be an issue with NaN handling. I don't reproduce it locally but I've seen a failure in CI after one of the pushes with `kl=ku=0`: the test expected a NaN but got a 0 (I think). Looking at `f64.L1NormInc`, I think it boils down to how MAXPD and MAXSD behave. https://www.felixcloutier.com/x86/maxsd says:

> If only one value is a NaN (SNaN or QNaN) for this instruction, the second source operand, either a NaN or a valid floating-point value, is written to the result.

That is, NaN may not be propagated as expected.

Reply:
Ok, I understand. The failure is in noasm build when `ldab=1`, then `inc=0` and the for loop in the stub doesn't execute and 0 is returned. So either pass `max(1,ldab-1)` (stride should be always at least 1) or change the stubs (the asm and noasm versions should ideally behave identically).

Reply:
I think it's all clear now, can you please re-approve? I'll add the benchmark later.

Reply:
Nice work. While I was looking into I found a nasty bug in extreme.